hibernate-core/src/main/java/org/hibernate/resource/beans/internal/JpaCdiLifecycleManagementStrategy.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.resource.beans.internal;
+
+import java.util.Set;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.inject.spi.AnnotatedType;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.InjectionTarget;
+
+import org.hibernate.resource.beans.spi.ManagedBean;
+
+/**
+ * A {@link CdiLifecycleManagementStrategy} to use when JPA compliance is required
+ * (i.e. when the bean lifecycle is to be managed by the JPA runtime, not the CDI runtime).
+ *
+ * The main characteristic of this strategy is that each requested bean is instantiated directly
+ * and guaranteed to not be shared in the CDI context.
+ *
+ * In particular, @Singleton-scoped or @ApplicationScoped beans are instantiated directly by this strategy,
+ * even if there is already an instance in the CDI context.
+ * This means singletons are not really singletons, but this seems to be the behavior required by
+ * the JPA 2.2 spec.
+ */
```

Comment:
But see, this is not true.  For a given class we only ever generate one ManagedBean instance = they are singletons (unless of course they are used again in a non-JPA context).  Notice the caching of them initially in a Map keyed by class...

Reply:
> unless of course they are used again in a non-JPA context

My point exactly. They are singleton in the JPA context only, not in the CDI context overall. I can rephrase the comment to clarify that if you like.

Reply:
Also, even in the JPA context, such "not-actually-singleton" beans will be instantiated once per `ServiceRegistry`, which unless I'm mistaken means once per `SessionFactory`/persistence unit. So even from the JPA point of view, they cannot always be viewed as singletons.